compiler_opt/rl/corpus.py

```diff
@@ -32,11 +34,128 @@ class ModuleSpec:
+
+  option = next(option_iterator, None)
+  cmdline = []
+  while option is not None:
+    if any(option.startswith(flag) for flag in delete_flags):
+      if '=' not in option:
+        next(option_iterator, None)
+    else:
+      matching_replace = [
+          flag for flag in replace_flags if option.startswith(flag)
+      ]
+      if not matching_replace:
+        cmdline.append(option)
+      else:
+        assert len(matching_replace) == 1
+        flag = matching_replace[0]
+        if flag in matched_replace_flags:
+          raise ValueError(f'{flag} was matched twice')
+        matched_replace_flags.add(flag)
+
+        if '=' not in option:
+          next(option_iterator, None)
+          cmdline.extend([option, replace_flags[flag].format(context=context)])
+        else:
+          cmdline.append(flag + '=' +
+                         replace_flags[flag].format(context=context))
+
+    option = next(option_iterator, None)
+  if len(matched_replace_flags) != len(replace_flags):
+    raise ValueError('flags that were expected to be replaced were not found')
+  cmdline.extend([f.format(context=context) for f in additional_flags])
+  return cmdline
+
+
+@dataclass(frozen=True)
+class LoadedModuleSpec:
+  """Encapsulates the loaded data of a module and the rules to persist it.
+
+  A LoadedModuleSpec can be passed to a remote location. There, given a local
+  directory, to_module_spec can be called, resulting in the data being saved
+  under that directory, the final compiler command line fully computed, and a
+  ready-to-use ModuleSpec returned.
+  """
+  name: str
+  module: bytes
+  thinlto_index: Optional[bytes] = None
+  orig_options: Tuple[str, ...] = ()
+  additional_flags: Tuple[str, ...] = ()
+  delete_flags: Tuple[str, ...] = ()
+  replace_flags: Optional[Dict[str, str]] = None
+
+  def _create_files_and_get_context(self, local_dir: str):
+    root_dir = os.path.join(local_dir, self.name)
+    os.makedirs(root_dir, exist_ok=True)
+    module_path = os.path.join(root_dir, 'input.bc')
+    thinlto_index_path = None
+    with tf.io.gfile.GFile(module_path, 'wb') as f:
+      f.write(self.module)
+    if self.thinlto_index is not None:
+      thinlto_index_path = os.path.join(root_dir, 'index.thinlto.bc')
+      with tf.io.gfile.GFile(thinlto_index_path, 'wb') as f:
+        f.write(self.thinlto_index)
+    context = Corpus.ReplaceContext(
+        module_full_path=module_path, thinlto_full_path=thinlto_index_path)
+    return context
+
+  def to_module_spec(self, local_dir: str) -> ModuleSpec:
+    context = self._create_files_and_get_context(local_dir)
+    cmdline = _localize_cmdline(
+        context=context,
+        orig_options=self.orig_options,
+        additional_flags=self.additional_flags,
+        delete_flags=self.delete_flags,
+        replace_flags=self.replace_flags)
+    return ModuleSpec(
+        name=self.name, exec_cmd=tuple(cmdline), size=len(self.module))
+
+
+@dataclass(frozen=True)
+class CorpusElement:
```

Comment:
It's in the intent. The `ModuleSpec` is basically a ready to use command line, everything has been prepared and ready to use. `CorpusElement` is tied to the `Corpus` object, it's very much not ready to use, the flag converters need to be run over it.

We could, actually, reduce the concept count to 1, and to avoid confusion I'll use this patch's terminology:
- remove `ModuleSpec` and replace with just the `Tuple[str, ...]` that's the fully resolved command line args
- `s/CorpusElement/ModuleSpec`

Basically, in the patch, the extra flags of ModuleSpec are not used, except for `name` which is used to raise an error. The error can be handled on the training side, who knows which corpus element this error pertains to (incl. to build up a repro script, out of the scope of this patch. But i'd just use the command line right now, with this proposal).

wdyt?

Reply:
Resolving because later discussion tackles this.